### PR TITLE
_getArtifact does not fetch all versions

### PR DESCRIPTION
**Pull request by dang-gyg**
currently, _getArtifact fetches all versions of an artifact, then checks for the first one with valid state. This has 2 problems:
- fetching all versions is expensive as all blobs are also fetched
- additional metadata query for each version until one with valid state is found

This PR optimizes the _getArtifact by performing at most only 1 query

### CODE REVIEW

**Review by carlesarnal — CHANGES_REQUESTED** [review]
Tests are not passing, and, according to what I see in the logs there is a syntax error in the query.

File: storage/jpa/src/main/java/io/apicurio/registry/storage/impl/jpa/JPARegistryStorage.java [hunk 1]
```diff
@@ -170,22 +170,41 @@ private Artifact _getArtifact(String artifactId, long version, EnumSet<ArtifactS
...
+                    .stream()
+                    .map(s -> s.toString())
+                    .collect(Collectors.joining(","));
+            String query = String.format(
+                     "SELECT a.* " +
+                     "FROM artifacts a " +
+                     "INNER JOIN " +
+                     "  (SELECT artifact_id, max(version) AS MaxVersion " +
+                     "  FROM meta " +
+                     "  WHERE artifact_id = '%s' AND key = '%s' AND value = ANY('{%s}'::text[]) " +
```
Comment (dmvolod) [inline storage/jpa/src/main/java/io/apicurio/registry/storage/impl/jpa/JPARegistryStorage.java:185]: No `ANY` operand is present on H2 database, that is the reason why tests are fails on CI
However, after changing to IN, the tests are still fails due to some logic errors

Comment (dmvolod) [inline storage/jpa/src/main/java/io/apicurio/registry/storage/impl/jpa/JPARegistryStorage.java:185]:
```suggestion
                     "  WHERE artifact_id = '%s' AND key = '%s' AND value IN ('{%s}') " +
```

Comment (dang-gyg) [reply storage/jpa/src/main/java/io/apicurio/registry/storage/impl/jpa/JPARegistryStorage.java]: thanks for the suggestion. I need to modify it a bit

**Comment by dang-gyg** [comment]
ArtifactStateTest test is still failing. I will look into it

**Comment by dmvolod** [comment]
> ArtifactStateTest test is still failing. I will look into it

Looks like it's a CI issue, I ran all tests locally and they were passing without errors

**Comment by dang-gyg** [comment]
I see. any hints how to fix/unblock this CI issue?

**Comment by dmvolod** [comment]
> I see. any hints how to fix/unblock this CI issue?

I'm not sure, but looks like this branch is out-of-date with the base branch, please rebase on the latest master and retry to test

**Comment by dmvolod** [comment]
@dang-gyg during some tests and enabling persistence for H2 found situation when max(version) inside ARTIFACTS table is `2` but in META is `3`. Looks like some inconsistent behavior is present. Need to be investigate. The previous logic works fine with this situation as don't care about max(version) in the META table and them consistency.

**Comment by dang-gyg** [comment]
@dmvolod could it be that the test is wrong? because even with the #819, the `max(version)` query considers only versions with valid states, so if version 3's state is `DELETED`, version 2's state is `ENABLED`, max(version) should correctly return 2, even when version 3's meta is still around.

**Comment by dmvolod** [comment]
> @dmvolod could it be that the test is wrong? because even with the #819, the `max(version)` query considers only versions with valid states, so if version 3's state is `DELETED`, version 2's state is `ENABLED`, max(version) should correctly return 2, even when version 3's meta is still around.

No, @dang-gyg . DELETED is not the state. This is physical deletion in JPA. The tests are correct and we just to fix #819 before this PR

**Comment by dmvolod** [comment]
@dang-gyg PR #822 has merged, please rebase on master again

**Comment by EricWittmann** [comment]
Hey guys - I appreciate the work you're doing in this area, but I wanted to let you know that we are evaluating a re-implementation of the JPA layer with a somewhat substantial change to the data model.  The JPA storage layer was implemented early in the project's evolution, and the data model chosen at that time didn't account for some of the features we eventually added (artifact search for example).  So you may want to hold off on too much more work in this area until we figure out what we're doing here.

If you'd like to have a look at some ongoing work in progress (proof of concept) in this area go here:

https://github.com/Apicurio/apicurio-registry/tree/features/sql

**Comment by dmvolod** [comment]
Yes, @EricWittmann sure. Maybe I missed something, but it would be nice to see some roadmap and under development features in the issues list.

**Comment by dang-gyg** [comment]
the build is green now. do you think it still makes sense to merge this change, while waiting for the new JPA layer?

**Comment by EricWittmann** [comment]
Here is a work in progress GitHub Project that may help: https://github.com/Apicurio/apicurio-registry/projects/1

It's not a proper roadmap, but it's *something*.  Unfortunately we are victims of constantly shifting priorities, which makes it challenging to put together and maintain a roadmap.  Sorry for that.

**Comment by EricWittmann** [comment]
This looks like a good change to me - now that the build is green I am merging it.

**Pull request merged**